### DISCUSSION TITLE
Strongswan pki tool libtls - x.509 based client authentication

### DISCUSSION THREAD

**Discussion by vojtechrichter23** [strongswan/strongswan#1784] [Q&A] [answered]
Hello,

I am trying to use strongswan pki tool for interaction with EST server. EST Server requires cert based authentication but in libtls library the client authentication is not allowed (also did not find anything in TLS options https://docs.strongswan.org/docs/5.9/config/tlsOptions.html) Does libtls in strongswan support cert based client authentication? Is there configuration file where can i enable it?

root@*:# pki --estca --url https://testrfc7030.com:9443 --caout testrfccert.pem --cacert dstcax3.pem
TPM 2.0 - could not load "libtss2-tcti-tabrmd.so.0"
plugin 'tpm': failed to load - tpm_plugin_create returned NULL
negotiated TLS 1.2 using suite TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
received TLS server certificate 'CN=testrfc7030.com'
received TLS intermediate certificate 'CN=estExampleCA'
  using certificate "CN=testrfc7030.com"
  using trusted ca certificate "CN=estExampleCA"
  reached self-signed root ca with a path length of 0
checking certificate status of "CN=testrfc7030.com"
certificate status is not available
server requested a certificate, but client authentication disabled
sending TLS close notify
L2 - certificate:  length of ASN.1 object invalid or too large
building CRED_CONTAINER - PKCS7 failed, tried 3 builders
  file coded in unknown format, discarded
building CRED_CERTIFICATE - X509 failed, tried 5 builders
could not parse single CA certificate

**Comment by tobiasbrunner** [comment]
Requiring client authentication for the `/cacerts` endpoint (which is what `--estca` triggers) doesn't really make sense to me. How would you bootstrap a client? [The RFC explicitly states](https://datatracker.ietf.org/doc/html/rfc7030#section-4.1.2): "The EST server SHOULD NOT require client authentication or authorization to reply to this request."

**Reply by vojtechrichter23** [reply]
Hello Tobias,

thanks for reply. As for /cacerts operation, the RFC do not exactly specifies if the TLS session must or must not be authenticated (RFC use should keyword ... yeah i know ... word plays). But it is common that certificate of the EST server TLS function (TLS Auth) differs from the one it is mediate issuing (EST CA), commonly for this certificate chain you must obtain the certificate chain from unencrypted/unauthenticated API for example some http URI. Then you initiate /cacert operation and you authenticate the EST TLS Session by the chain you obtained offline/from unauthenticated API and also get cert-chain of EST CA tree ... Also all EST services should be authenticated, because also server needs to know if the client is worthy for his services - yes /cacerts is only returning public certificates so it is not dangerous operation from security perspective ...

But /cacerts operation does not matter that much for me i can accept also unauthenticated one (option for authenticated one could be nice) ..., but when i use pki --est (which should invoke /simpleenroll for first-time cert enroll or /simplereenroll operation for renew-cert) i get same error and could not find the option for enabling EST Client authentication in strongswan for libtls library (could not find right TLS options or some setting in charon.conf / dependents module - *.conf ...) can you help me with this?:

root@deb-12-est:/home/estlab# pki --est --url https://testrfc7030.com:9443 --in rfc7030-self-req.pem --cacert dstcax3.pem --outform pem > rfc7030-signed-cert.pem
TPM 2.0 - could not load "libtss2-tcti-tabrmd.so.0"
plugin 'tpm': failed to load - tpm_plugin_create returned NULL
negotiated TLS 1.2 using suite TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
received TLS server certificate 'CN=testrfc7030.com'
received TLS intermediate certificate 'CN=estExampleCA'
  using certificate "CN=testrfc7030.com"
  using trusted ca certificate "CN=estExampleCA"
  reached self-signed root ca with a path length of 0
checking certificate status of "CN=testrfc7030.com"
certificate status is not available
server requested a certificate, but client authentication disabled
sending TLS close notify
EST request failed: HTTP 401

**Reply by tobiasbrunner** [reply]
> (RFC use should keyword ... yeah i know ... word plays).

No, not wordplays, SHOULD NOT is not the same as should not in RFCs.

> But it is common that certificate of the EST server TLS function (TLS Auth) differs from the one it is mediate issuing (EST CA), commonly for this certificate chain you must obtain the certificate chain from unencrypted/unauthenticated API for example some http URI.

I'm not sure I fully understand what you mean. Could you please point to the parts of the RFC where this is described.

> Also all EST services should be authenticated, because also server needs to know if the client is worthy for his services

The RFC says that's not the case. For instance in section 2.1 about the `/cacert` operation: "Client authentication is not required for this exchange, so it is trivially supported by the EST server"


> but when i use pki --est (which should invoke /simpleenroll for first-time cert enroll or /simplereenroll operation for renew-cert) i get same error

That's because it only supports certificate authentication for re-enrollment (you can pass the previous cert and key via `--cert` and `--key`). For the initial enrollment it only supports HTTP basic authentication (via `--userpass`).

**Reply by vojtechrichter23** [reply]
No, not wordplays, SHOULD NOT is not the same as should not in RFCs.

    I am not joining this fight. As i wrote i dont mind, that in the --estca function for /cacerts call there is no client authentication.

I'm not sure I fully understand what you mean. Could you please point to the parts of the RFC where this is described.
    
    Section 3. EST server certificate -> A CA authenticatable by a third-party TA e.g., a web server CA

    EST Server is using certificate from different CA Chain for TLS authentication (other than for EST CA that is serving - in /cacert message
    there is EST CA Chain, but when EST Server is using certificate issued by 3rd party CA, then you need other way to obtain EST Server 
    cert/trust chain - for example in factory to install to explicit TA, by API or implicit Trust like in Bootstrap section 4.1.1 ) Its also easier to 
    expose HTTP API for EST Server Cert chain than programming Bostrap behaviour (implicit vs explicit TA and OOB cert auth)

That's because it only supports certificate authentication for re-enrollment (you can pass the previous cert and key via --cert and --key). For the initial enrollment it only supports HTTP basic authentication (via --userpass).
    
     Thats the information i needed. In our EST implementation we are planning to use manufacturer-installed cert (for example DEVICE-ID cert  
     in TAM) for frist time authentication to EST Server described in section 2.2 "TLS with a previously installed certificate (e.g., manufacturer-
     installed certificate or a certificate issued by some other party), because its more secure than HTTP Basic authentication (desribed in section 
     6. Security Considerations) and it also scales more (thousands of devices) - when u use HTTP Basic there are 2. options:
     1. all requests are authenticated by same user/pass combination or Digest - that is not really secure (same as in SCEP challenge password 
     in CSR)
     2- all requests use differnet user/pass or Digest combination - then u need some other trusted system to distribute this information (and 
     you need to authenticate the end-system - ussually by manufacturer certificate)

To summarize the question, we are using strongswan for large scale routed based VPNs, now with SCEP enrollment protocol for ensuring auto-enroll functions for certificates which we are using for IKEv2 Auth phase authentication ... We are now migrating to EST (we have combination of Linux based devices and cisco devices ...) and i am researching if EST client in strogswan would be capable of working in our EST sceneario ... which need to support both clients (linux and cisco - which is hardcoded) - EST Server is authenticating clients by their manufacturers certificates (he is trusting manufacturers CAs as described in Explicit TA) and it also looks in certificate for SerialNumber (on EST Server / CA is list with SN -> allowed devices which can request certificates from EST Server / CA)

**Reply by tobiasbrunner** [reply, marked as answer]
I see. That would require some changes to the EST client as it currently automatically uses `/simplereenroll` when a certificate and key is passed. So that would require maybe a new option to force a `/simpleenroll`.
